pinot-core/src/main/java/org/apache/pinot/server/realtime/ControllerLeaderLocator.java

```diff
@@ -80,45 +84,111 @@ public static ControllerLeaderLocator getInstance() {
+   * Otherwise, updates lead controller pairs from Helix cluster leader.
+   * @return true if refresh is completed successfully.
    */
-  private Pair<String, Integer> getLeaderForTable(String rawTableName) {
+  private boolean refreshControllerLeaderMap() {
     // Checks whether lead controller resource has been enabled or not.
     if (isLeadControllerResourceEnabled()) {
-      // Gets leader from lead controller resource.
-      return getLeaderFromLeadControllerResource(rawTableName);
+      return refreshControllerLeaderMapFromLeadControllerResource();
     } else {
-      // Gets Helix leader to be the leader to this table, otherwise returns null.
-      return getHelixClusterLeader();
+      return refreshControllerLeaderMapFromHelixClusterLeader();
+    }
+  }
+
+  /**
+   * Updates lead controller pairs from the external view of lead controller resource.
+   * @return true if refresh is completed successfully, false if no leader found or exception thrown.
+   */
+  private boolean refreshControllerLeaderMapFromLeadControllerResource() {
+    try {
+      ExternalView leadControllerResourceExternalView = _helixManager.getClusterManagmentTool()
+          .getResourceExternalView(_helixManager.getClusterName(), CommonConstants.Helix.LEAD_CONTROLLER_RESOURCE_NAME);
+      if (leadControllerResourceExternalView == null) {
+        LOGGER.warn("External view of lead controller resource is null.");
+        return false;
+      }
+      Set<String> partitionNames = leadControllerResourceExternalView.getPartitionSet();
+      if (partitionNames.isEmpty()) {
+        LOGGER.warn("The partition set of lead controller resource is empty.");
+        return false;
+      }
+      if (partitionNames.size() != CommonConstants.Helix.NUMBER_OF_PARTITIONS_IN_LEAD_CONTROLLER_RESOURCE) {
+        LOGGER.warn("The partition size of lead controller resource isn't {}. Actual size: {}",
+            CommonConstants.Helix.NUMBER_OF_PARTITIONS_IN_LEAD_CONTROLLER_RESOURCE, partitionNames.size());
+        return false;
+      }
+      for (String partitionName : partitionNames) {
+        int partitionId = LeadControllerUtils.extractPartitionId(partitionName);
+        Map<String, String> partitionStateMap = leadControllerResourceExternalView.getStateMap(partitionName);
+        boolean masterFound = false;
+        // Get master host from partition map. Return null if no master found.
+        for (Map.Entry<String, String> entry : partitionStateMap.entrySet()) {
+          if (MasterSlaveSMD.States.MASTER.name().equals(entry.getValue())) {
+            // Found the controller in master state.
+            // Converts participant id (with Prefix "Controller_") to controller id and assigns it as the leader,
+            // since realtime segment completion protocol doesn't need the prefix in controller instance id.
+            String participantInstanceId = entry.getKey();
```

Comment:
Didnt we agree to move the encode/decode of controller instanceId to participant instance Id to LeadControllerUtils?

Reply:
No, in this PR only Pinot server needs to fetch the external view.